docs/source/daml/intro/8_Upgrading.rst

```diff
@@ -0,0 +1,170 @@
+
+Executing an Upgrade
+--------------------
+
+The ``Test`` modules of the ``8_Upgrade`` project demonstrate the upgrade using DAML Script. 
+
+You'll notice a module ``Test.Intro.Asset.Upgrade.V1Setup``, which is almost a carbon copy of the V1 trade setup Scripts. ``data-dependencies`` is designed to use existing contracts and data types. DAML Script is not imported. In practice, we also shouldn't expect that the DAR file we download from the ledger using ``daml ledger fetch-dar`` contains test scripts. For larger projects it's good practice to keep them separate and only deploy templates to the ledger.
+
+The entry point for the upgrade is the script ``Test.Intro.Asset.Upgrade.V1Setup.testUpgrade``. The ledger is set up using the V1 trade setup scripts. The four relationships between banks and asset holders are then captured in a ``Relationship`` type, and then executed using a ``forA_`` statement. We'll revisit ``forA`` and its underscored version later. For now, read ``forA_ xs fn`` as "For each ``x`` in ``xs``, perform the action ``fn x``". The action in question here is ``runCompleteUpgrade``, which as the name suggests, outlines a complete upgrade for the relationship ``r``:
+
+.. literalinclude:: daml/daml-intro-8-upgrade/daml/Test/Intro/Asset/Upgrade/V2.daml
+  :language: daml
+  :start-after: -- RUN_COMPLETE_UPGRADE_BEGIN
+  :end-before: -- RUN_COMPLETE_UPGRADE_END
+
+We won't go in depth into each of the steps in this chapter, but here are the steps performed, in line with the "Anatomy" above:
+
+#. In ``initiateUpgrade`` the ``issuer`` of the upgrade checks for V1 ``AssetHolder`` and  ``AssetHolderInvite``, and creates corresponding ``UpgradeInvite`` or V2  ``AssetHolderInvite``.
+#. In ``acceptUpgrade`` the ``owner`` matches up V1 ``AssetHolder`` contracts with ``UpgradeInvite`` contracts, and calls the ``Accept_Upgrade`` choice for each pair.
+#. In ``performUpgrade`` the ``owner`` tidies up by upgrading all V1 ``Asset`` contracts and cancelling all ``TransferApproval`` and ``TransferProposal`` contracts.
+#. In ``ownerCheckUpgradeComplete`` the ``owner`` checks whether they think the upgrade for the given relationship is complete.
+#. In ``confirmCompletion`` the ``owner`` creates the ``UpgradeConfirmation`` contract for each ``Upgrade`` contract.
+#. In ``issuerCheckUpgradeComplete`` the ``issuer`` checks whether they think the upgrade is complete.
+#. In ``completeUpgrade`` the ``issuer`` matches up ``Upgrade`` with ``UpgradeConfirmation`` contracts and completes the process.
+
+In a soon to come Chapter 10 of the introduction to DAML you'll learn why the Scripts are structured the way they are, and how to use them to test beyond the IDE.
+
+This concludes this introduction to upgrades. If you want to learn more, you can also check out the standalone docs pages on upgrading: :doc:`/upgrade/index`.
+
+Next up
```

Comment:
I don’t understand the purpose of this chapter. We already have docs on upgrading, if they are insufficient, why not improve them instead of adding a completely new chapter which describes the same thing? I don’t think this is any more introductory than the upgrade docs (quite the opposite if you want to keep all the details on interface files and similar stuff).

Reply:
I understand what you are saying, and I'm not particularly happy with having two sections on upgrading in our docs, either. My aim with the "Intro to DAML" as a whole is to have a single "tutorial" style body of docs that gets devs to the point where they can take the DAML contract certification. Upgrading specifically isn't covered in there, but dependencies, data-dependencies, and to some extent architecture are - identifiers, what DAML-LF is, etc.

Furthermore, the Intro to DAML up to this point is structured as developing a single application - an asset holding model. I couldn't think of a better example to explain the above concepts in this context than to talk about upgrading. Maybe I should just have separated it out into separate projects for data, rules and tests, or something similar.